app/Models/Course.php

```diff
@@ -18,14 +18,68 @@ class Course extends Model
         'description'
     ];
 
+    public function scopeSearch($query, $data)
+    {
+        if (isset($data['keyword'])) {
+            $query->where('name', 'LIKE', "%". $data['keyword'] ."%")->orWhere('description', 'LIKE', "%". $data['keyword'] ."%");
+        }
+
+        if (isset($data['created_time'])) {
+            $query->orderBy('courses.created_at', $data['created_time']);
+        }
+
+        if (isset($data['teacher_id']) && !empty($data['teacher_id'])) {
+            $query->join('teacher_course', 'courses.id', '=', 'teacher_course.course_id')
+                ->where('user_id', $data['teacher_id']);
+        }
+
+        if (isset($data['learner']) && !empty($data['learner'])) {
+            $query->withCount('users')->orderBy('users_count', $data['learner']);
+        }
+
+        if (isset($data['time']) && !empty($data['time'])) {
+            $query->withCount('lessons')->withSum('lessons', 'times')->orderBy('lessons_sum_times', $data['time']);
+        }
+
+        if (isset($data['lesson']) && !empty($data['lesson'])) {
+            $query->withCount('lessons')->orderBy('lessons_count', $data['lesson']);
+        }
+
+        if (isset($data['tags']) && count($data['tags']) > 0) {
+            $query->join('course_tag', 'courses.id', '=', 'course_tag.course_id')
+                ->whereIn('tag_id', $data['tags']);
+        }
+
+        if (isset($data['rate']) && !empty($data['rate'])) {
+            $query->withCount('reviews')->orderBy('reviews_count', $data['rate']);
```

Comment:
câu này hiểu như thế nào

Reply:
select `courses`.*, 
(select count(*) from `reviews` where `courses`.`id` = `reviews`.`course_id` and `reviews`.`deleted_at` is null) as `reviews_count` 
from `courses` 
order by `reviews_count` desc

thực hiện nối bảng courses với bảng reviews,
trả về các bản ghi gồm tất cả các trường trong bảng course, trường reviews_count (là đếm số reviews theo course_id)
được sắp xếp theo reviews_count